doc/neps/nep-0040-legacy-datatype-impl.rst

```diff
@@ -0,0 +1,616 @@
+================================================
+NEP 40 — Legacy Datatype Implementation in NumPy
+================================================
+
+:title: Legacy Datatype Implementation in NumPy
+:Author: Sebastian Berg
+:Status: Draft
+:Type: Informational
+:Created: 2019-07-17
+
+
+Abstract
+--------
+
+As a preparation to further NumPy enhancement proposals 41, 42, and 43. This
+NEP details the current status of NumPy datatypes as of NumPy 1.18.
+It describes some of the technical aspects and concepts that
+motivated the other proposals.
+
+
+Detailed Description
+--------------------
+
+This section describes some central concepts and provides a brief overview
+of the current implementation of dtypes as well as a discussion.
+In many cases subsections will be split roughly to first describe the
+current implementation and then follow with an "Issues and Discussion" section.
+
+Parametric Datatypes
+^^^^^^^^^^^^^^^^^^^^
+
+Some datatypes are inherently *parametric*. All ``np.flexible`` scalar
+types are attached to parametric datatypes (string, bytes, and void).
+The class ``np.flexible`` for scalars is a superclass for the data types of
+variable length (string, bytes,
+and void). This distinction is similarly exposed by the C-Macros
+``PyDataType_ISFLEXIBLE`` and ``PyTypeNum_ISFLEXIBLE``.
+For instance, ``"S8"`` can represent longer strings than ``"S4"``.
```

Comment:
This sentence feels a little out of place here, but probably is sensible to work in somehow.